USF-1327: Checkout containers (redux)

jeff-matthews opened the pull request:
This pull request (PR) documents Checkout container configuration options based on the [doc template](https://github.com/commerce-docs/microsite-commerce-storefront/blob/develop/_dropin-templates/dropin-containers.mdx?plain=1) for dropin containers.

Options and descriptions are based on the prop interfaces defined for each component in the source code.

jeff-matthews committed:
USF-1327: Initial draft of checkout container docs

jeff-matthews committed:
USF-1327: Rename checkout files for simplicity

jeff-matthews committed:
USF-1327: Separate checkout containers into separate files

jeff-matthews committed:
USF-1327: Add description metadata

mcocapiza commented on src/content/docs/dropins/checkout/containers/estimate-shipping.mdx:
```diff
@@ -0,0 +1,24 @@
+    ['taxExcluded', 'boolean', 'No', 'A boolean indicating whether the tax is excluded from the shipping price.'],
+    ['taxIncluded', 'boolean', 'No', 'A boolean indicating whether the tax is included in the shipping price.'],
+  ]}
+/>
```
Although `EstimateShipping` is a checkout container, it's a bit special. We had to create this one to overwrite the `EstimateShipping` defined into the Storefront Cart dropin. Whereas the `EstimateShipping` defined into the Storefront Cart dropin is editable, the `EstimateShipping` defined into the Checkout dropin is readonly.
Initially it shows the estimate shipping costs, but once the customer has introduced a shipping address and a shipping method is selected, then it shows the real shipping cost.
You can see the estimate shipping information inside of the `OrderSummary` container (comming from the Storefront Cart dropin)... so it only makes sense to use it as slot of the `OrderSummary` container.

**example**:

```
import { OrderSummary } from '@dropins/storefront-cart/containers/OrderSummary.js';
import { render as CartProvider } from '@dropins/storefront-cart/render.js';

const $orderSummary = checkoutFragment.querySelector(
  '.checkout__order-summary'
);

CartProvider.render(OrderSummary, {
  slots: {
    EstimateShipping: (esCtx) => {
      const estimateShippingForm = document.createElement('div');
      CheckoutProvider.render(EstimateShipping)(estimateShippingForm);
      esCtx.appendChild(estimateShippingForm);
    },
  },
})($orderSummary),
```

mcocapiza commented on src/content/docs/dropins/checkout/containers/login-form.mdx:
```diff
@@ -0,0 +1,28 @@
+    ['onSignOutClick', 'function', 'No', 'A function that handles the sign-out button click. It takes no arguments.'],
+    ['customer', 'Customer | null', 'No', 'The customer information, if available.']
+  ]}
+/>
```
`LoginForm` uses the `AuthCombine` container defined into the Storefront Auth dropin.

**example**:

```
import AuthCombine from '@dropins/storefront-auth/containers/AuthCombine.js';
import { render as AuthProvider } from '@dropins/storefront-auth/render.js';

import LoginForm from '@dropins/storefront-checkout/containers/LoginForm.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $login = checkoutFragment.querySelector('.checkout__login');

CheckoutProvider.render(LoginForm, {
  name: LOGIN_FORM_NAME,
  onSignInClick: async (initialEmailValue) => {
    const signInForm = document.createElement('div');

    AuthProvider.render(AuthCombine, {
      signInFormConfig: {
        renderSignUpLink: true,
        initialEmailValue,
        onSuccessCallback: () => {
          displayOverlaySpinner();
        },
      },
      signUpFormConfig: {},
      resetPasswordFormConfig: {},
    })(signInForm);

    showModal(signInForm);
  },
  onSignOutClick: () => {
    authApi.revokeCustomerToken();
  },
})($login),
```

mcocapiza commented on src/content/docs/dropins/checkout/containers/out-of-stock.mdx:
```diff
@@ -0,0 +1,22 @@
+    ['onRemoveOutOfStock', 'function', 'No', 'Handles the removal of out-of-stock items. It takes an event as an argument.'],
+    ['routeCart', 'string', 'No', 'The route to the cart page.']
+  ]}
+/>
```
**example**:

```
import OutOfStock from '@dropins/storefront-checkout/containers/OutOfStock.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $outOfStock = checkoutFragment.querySelector('.checkout__out-of-stock');

CheckoutProvider.render(OutOfStock, {
  routeCart: () => '/cart',
  onCartProductsUpdate: (items) => {
    cartApi.updateProductsFromCart(items).catch(console.error);
  },
})($outOfStock),
```

mcocapiza commented on src/content/docs/dropins/checkout/containers/payment-methods.mdx:
```diff
@@ -0,0 +1,21 @@
+    ['slots.Main', 'SlotProps<PaymentMethodsMainSlotContext>', 'No', 'The main slot for the payment methods.'],
+    ['slots.Handlers', 'PaymentMethodHandlerSlots', 'No', 'The handlers for different payment methods.']
+  ]}
+/>
```
**example**:

```
import PaymentMethods from '@dropins/storefront-checkout/containers/PaymentMethods.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $paymentMethods = checkoutFragment.querySelector(
  '.checkout__payment-methods'
);

CheckoutProvider.render(PaymentMethods)($paymentMethods),
```

mcocapiza commented on src/content/docs/dropins/checkout/containers/place-order.mdx:
```diff
@@ -0,0 +1,21 @@
+    ['onPlaceOrder', 'function', 'No', 'Handles the place order action. It takes a PlaceOrderContext object as an argument and returns a Promise<void>.'],
+    ['disabled', 'boolean', 'No', 'Indicates whether the place order button is disabled.']
+  ]}
+/>
```
`PlaceOrder` container accepts a function as a prop to be used to perform some validations (like forms, etc):

```handleValidation?: () => boolean;```

**example**:

```
import PlaceOrder from '@dropins/storefront-checkout/containers/PlaceOrder.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $placeOrder = checkoutFragment.querySelector('.checkout__place-order');

CheckoutProvider.render(PlaceOrder, {
  handleValidation: () => {
    let success = true;
    const forms = document.forms;

    const loginForm = forms[LOGIN_FORM_NAME];

    if (loginForm) {
      success = loginForm.checkValidity();
      if (!success) scrollToElement($login);
    }

    const shippingForm = forms[SHIPPING_FORM_NAME];

    if (
      success &&
      shippingFormRef.current &&
      shippingForm &&
      shippingForm.checkVisibility()
    ) {
      success = shippingFormRef.current.handleValidationSubmit(false);
    }

    const billingForm = forms[BILLING_FORM_NAME];

    if (
      success &&
      billingFormRef.current &&
      billingForm &&
      billingForm.checkVisibility()
    ) {
      success = billingFormRef.current.handleValidationSubmit(false);
    }

    return success;
  },
  onPlaceOrder: async () => {
    displayOverlaySpinner();

    try {
      await checkoutApi.placeOrder();
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      removeOverlaySpinner();
    }
  },
})($placeOrder),
]);
```

mcocapiza commented on src/content/docs/dropins/checkout/containers/server-error.mdx:
```diff
@@ -0,0 +1,22 @@
+    ['contactSupport', 'string', 'No', 'Provides a contact support message or link.'],
+    ['onClick', 'function', 'No', 'Handles click events. It takes an event (e: any) as an argument.']
+  ]}
+/>
```
**example**:

```
import ServerError from '@dropins/storefront-checkout/containers/ServerError.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $serverError = checkoutFragment.querySelector(
  '.checkout__server-error'
);

CheckoutProvider.render(ServerError, {
  onRetry: () => {
    $content.classList.remove('checkout__content--error');
  },
  onServerError: () => {
    $content.classList.add('checkout__content--error');
  },
})($serverError),
```

mcocapiza commented on src/content/docs/dropins/checkout/containers/shipping-methods.mdx:
```diff
@@ -0,0 +1,25 @@
+    ['className', 'string', 'No', 'The CSS class name(s) to apply to the component.'],
+    ['data-testid', 'string', 'No', 'Used for testing purposes to identify the component.']
+  ]}
+/>
```
**example**:

```
import ShippingMethods from '@dropins/storefront-checkout/containers/ShippingMethods.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $delivery = checkoutFragment.querySelector('.checkout__delivery');

CheckoutProvider.render(ShippingMethods, {
  hideOnVirtualCart: true,
  onCheckoutDataUpdate: () => {
    cartApi.refreshCart().catch(console.error);
  },
})($delivery),
```

mcocapiza reviewed (changes requested):
Also, you need to add a couple of more containers:
- BillToShippingAddress => Checkbox to decide if the billing address will be the same as the shipping address, otherwise the billing address form is shown

**example**:

```
import BillToShippingAddress from '@dropins/storefront-checkout/containers/BillToShippingAddress.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $billToShipping = checkoutFragment.querySelector(
  '.checkout__bill-to-shipping'
);

CheckoutProvider.render(BillToShippingAddress, {
  hideOnVirtualCart: true,
  onChange: (checked) => {
    $billingForm.style.display = checked ? 'none' : 'block';

    if (!checked && billingFormRef.current) {
      const isDataValid = billingFormRef.current.handleValidationSubmit();

      setAddressOnCart(
        { data: billingFormRef.current.formData, isDataValid },
        checkoutApi.setBillingAddress
      );
    }
  },
})($billToShipping),
```

- OrderConfirmationHeader => Once the order is placed, a new order confirmation page is shown with the order details. This container also has the "Create an account" button (Sign-up) in case the guest user wanted to create an account

**example**:

```
import OrderConfirmationHeader from '@dropins/storefront-checkout/containers/OrderConfirmationHeader.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $orderConfirmationHeader = orderConfirmationFragment.querySelector(
  '.order-confirmation__header'
);

CheckoutProvider.render(OrderConfirmationHeader, {
  orderData,
  onSignUpClick,
})($orderConfirmationHeader);
```

mcocapiza commented on src/content/docs/dropins/checkout/containers/error-banner.mdx:
```diff
@@ -0,0 +1,26 @@
+    ['onDismiss', 'function', 'No', 'Function to call when the banner is dismissed.'],
+    ['variant', 'string', 'No', 'Variant of the banner indicating the type of alert.'],
+  ]}
+/>
```
Please remove this block... we are not going to use it.

jeff-matthews committed:
USF-1327: Applied suggestions from tech review

jeff-matthews committed:
USF-1327: Added new checkout containers

jeff-matthews committed:
Merge branch 'USF-1323-checkout-dropin' into USF-1327-checkout-containers-redux

jeff-matthews committed:
USF-1327: Fixed validation error

jeff-matthews commented:
Thanks @mcocapiza. I addressed your feedback and added the additional containers (`BillToShippingAddress` and `OrderConfirmationHeader`).

Will you please provide a code sample for the `MergedCartBanner` container?

jeff-matthews committed:
Merge branch 'USF-1323-checkout-dropin' into USF-1327-checkout-containers-redux

jeff-matthews committed:
USF-1327: Added example for MergedCartBanner container

jeff-matthews commented:
I found a simple example in Slack. I'll push changes in a moment.

mcocapiza commented:
- `MergedCartBanner` => When the customer performs a "Sign-In" action, if he/she/it had a previous cart within products inserted, then a banner is shown informing the current items added into the guest cart and the items recovered from a previous customer cart will be merged. 

**example**:

```
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $mergedCartBanner = checkoutFragment.querySelector(
  '.checkout__merged-cart-banner',
);

CheckoutProvider.render(MergedCartBanner)($mergedCartBanner),
```

mcocapiza commented:
Please update the `EstimateShipping` example (I missed to import the `EstimateShipping` container from Checkout dropin):

```
import { OrderSummary } from '@dropins/storefront-cart/containers/OrderSummary.js';
import { render as CartProvider } from '@dropins/storefront-cart/render.js';
import EstimateShipping from '@dropins/storefront-checkout/containers/EstimateShipping.js';

const $orderSummary = checkoutFragment.querySelector(
  '.checkout__order-summary'
);

CartProvider.render(OrderSummary, {
  slots: {
    EstimateShipping: (esCtx) => {
      const estimateShippingForm = document.createElement('div');
      CheckoutProvider.render(EstimateShipping)(estimateShippingForm);
      esCtx.appendChild(estimateShippingForm);
    },
  },
})($orderSummary),

```

mcocapiza commented:
Please update the `OutOfStock` example (I missed to import the `cartApi` from Cart dropin):

```
import * as cartApi from '@dropins/storefront-cart/api.js';
import OutOfStock from '@dropins/storefront-checkout/containers/OutOfStock.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $outOfStock = checkoutFragment.querySelector('.checkout__out-of-stock');

CheckoutProvider.render(OutOfStock, {
  routeCart: () => '/cart',
  onCartProductsUpdate: (items) => {
    cartApi.updateProductsFromCart(items).catch(console.error);
  },
})($outOfStock),
```

mcocapiza commented:
Please update the `PlaceOrder` example (I missed to import the `checkoutApi` from Checkout dropin):

```
import * as checkoutApi from '@dropins/storefront-checkout/api.js';
import PlaceOrder from '@dropins/storefront-checkout/containers/PlaceOrder.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $placeOrder = checkoutFragment.querySelector('.checkout__place-order');

CheckoutProvider.render(PlaceOrder, {
  handleValidation: () => {
    let success = true;
    const forms = document.forms;

    const loginForm = forms[LOGIN_FORM_NAME];

    if (loginForm) {
      success = loginForm.checkValidity();
      if (!success) scrollToElement($login);
    }

    const shippingForm = forms[SHIPPING_FORM_NAME];

    if (
      success &&
      shippingFormRef.current &&
      shippingForm &&
      shippingForm.checkVisibility()
    ) {
      success = shippingFormRef.current.handleValidationSubmit(false);
    }

    const billingForm = forms[BILLING_FORM_NAME];

    if (
      success &&
      billingFormRef.current &&
      billingForm &&
      billingForm.checkVisibility()
    ) {
      success = billingFormRef.current.handleValidationSubmit(false);
    }

    return success;
  },
  onPlaceOrder: async () => {
    displayOverlaySpinner();

    try {
      await checkoutApi.placeOrder();
    } catch (error) {
      console.error(error);
      throw error;
    } finally {
      removeOverlaySpinner();
    }
  },
})($placeOrder),
]);
```

mcocapiza commented:
The `ServerError` configurations are the following ones:
- onRetry?: () => void; /* what to do in case of retry */
- onServerError?: (error: string) => void; /* what to do in case of a server error */

mcocapiza commented:
Please update the `ShippingMethods` example (I missed to import the `cartApi` from Cart dropin): 

```
import * as cartApi from '@dropins/storefront-cart/api.js';
import ShippingMethods from '@dropins/storefront-checkout/containers/ShippingMethods.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $delivery = checkoutFragment.querySelector('.checkout__delivery');

CheckoutProvider.render(ShippingMethods, {
  hideOnVirtualCart: true,
  onCheckoutDataUpdate: () => {
    cartApi.refreshCart().catch(console.error);
  },
})($delivery),
```

mcocapiza commented:
The `ShippingMethods` configurations are the following ones:
- onCheckoutDataUpdate?: () => Promise<void>; /* what to do in case the checkout data has been updated */
- onShippingMethodSelect?: (method: ShippingMethod) => void; /* what to do in case a shipping method has been selected */
- preSelectedMethod?: { carrierCode: string; methodCode: string }; /* if we want to preselect a specific shipping method */
- shippingMethodsSlot?: SlotProps; /* to provide more shipping methods */

Also we could include another prop:
- hideOnVirtualCart: boolean No Hides the container if the cart is virtual.

jeff-matthews committed:
USF-1327: Applied suggestions from technical review

jeff-matthews commented:
Thanks @mcocapiza. All revisions applied.

jeff-matthews committed:
USF-1327: Apply propoer naming convention to dropins

jeff-matthews committed:
Merge branch 'USF-1323-checkout-dropin' into USF-1327-checkout-containers-redux
